### PR TITLE
Support v1/chat/completions

### LINKED ISSUE #26: OpenAI Chat Completion Endpoint
First of all, thank you for such a great framework and study! Do you plan to support `chat/completions` endpoint as well for the models utilizing their `chat_template` for completion during backend serving?

### PR DESCRIPTION
**Pull request by comaniac**
Close #26 

- Support OpenAI compatible chat API with and without streaming.
- Support builtin and dynamic registered chat templates.

### CODE REVIEW

**Comment by merrymercy** [comment]
For the chat template, can we use the hugging face tokenizer by default? https://huggingface.co/docs/transformers/main/en/chat_templating#how-do-i-use-chat-templates

**Comment by comaniac** [comment]
> For the chat template, can we use the hugging face tokenizer by default? https://huggingface.co/docs/transformers/main/en/chat_templating#how-do-i-use-chat-templates

Good idea. Let me add this.

**Comment by comaniac** [comment]
Now if `--chat-template` is not specified, we use the tokenizer build-in chat template. Thus, most users should not worry about the chat template for HF models. Meanwhile, I found that some tokenizers such as TinyLlama does not have chat template, so we will get the following when calling `apply_chat_template(messages, tokenize=False, add_generation_prompt=True)`:

```
No chat template is defined for this tokenizer - using the default template for the LlamaTokenizerFast class. If the default is not appropriate for your model
, please set `tokenizer.chat_template` to an appropriate template. See https://huggingface.co/docs/transformers/main/chat_templating for more information.

'<s>[INST] <<SYS>>\nYou are a helpful AI assistant\n<</SYS>>\n\nList 3 countries and their capitals. [/INST]'
```

And this template is actually incorrect for this model, so you will get the following response in the unit test:

```
 <<SYS>>
You are a helpful AI assistant
<</SYS>>

List 4 cities with more than 100,000 people. [/INST] <<SYS>>
You are a helpful AI assistant
<</SYS>>

List 5
```

The following response is expected with ChatML template:

```
Here are three countries and their capitals:

1. Canada - Ottawa
2. Australia - Canberra
3. South Korea - Seoul
```

Another issue is `apply_chat_template` doesn't provide stop strings, so the response may be undesired. For example, Llama-2 template should have stop string `"[INST]", "[/INST]", "<<SYS>>", "<</SYS>>"`. In general, I would still suggest specify chat template explicitly, but maybe we could put this issue in the troubleshooting.

File: python/sglang/srt/conversation.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
+# Adapted from
+# https://github.com/lm-sys/FastChat/blob/main/fastchat/conversation.py
```
Comment (merrymercy) [inline python/sglang/srt/conversation.py:2]: We can consider importing instead of copying later.

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -77,6 +78,12 @@ def add_cli_args(parser: argparse.ArgumentParser):
             "tokenizer if available, and 'slow' will "
             "always use the slow tokenizer.",
         )
+        parser.add_argument(
+            "--chat-template",
+            type=str,
+            default=ServerArgs.chat_template,
+            help="The buliltin chat template name or the path of the chat template file.",
```
Comment (merrymercy) [inline python/sglang/srt/server_args.py]:
```suggestion
            help="The buliltin chat template name or the path of the chat template file. This is only used for OpenAI-compatible API server",
```

**Review by merrymercy — APPROVED** [review]

**Pull request merged**